### DISCUSSION TITLE
Retrieve corners of the stitched result

### DISCUSSION THREAD

**Discussion by neriniccolodei** [OpenStitching/stitching#149] [General]
Hello, 

I am trying to retrieve the corners of a stitched image, as I would like to stretch it out to make the corners coincide with the ones of a rectangle. This is an idea I had in order to minimise the miniaturisation effect observed when stitching together multiple frames in sequence, as displayed here:
![finer_orb_lc1](https://github.com/OpenStitching/stitching/assets/62433425/c411b617-f6be-4f76-ac8d-432d0b6077a6)

What I'd like to do, starting from a simple two-frames stitch, is retrieve the coordinates of the corners highlighted in red, here (note that this is another stitched image, no the simple one I'm working with, but just to understand what are "corners" to me):
![Picture 1](https://github.com/OpenStitching/stitching/assets/62433425/5028ab8f-d425-4c79-8d04-8e95a348c25c)

I have tried to follow the code and I have drawn the absolute value of `final_corners`, but the resulting corners are these, which are not what I'm looking for. 

![Corners](https://github.com/OpenStitching/stitching/assets/62433425/dcdf2cd4-d01f-4a5c-808e-66c54badd890)

Any idea on how to retrieve these?

Thank you and best regards.

**Comment by lukasalexanderweber** [comment]
I'm quite sure that you can get the points you need from the `final_sizes` and `final_corners` values. See [here](https://github.com/OpenStitching/stitching_tutorial/blob/master/docs/Stitching%20Tutorial.md#crop)

**Reply by lukasalexanderweber** [reply]
You might need to zero_center them, if I remember correctly. Maybe try to move along the cropping part

**Reply by lukasalexanderweber** [reply]
This is a follow up to https://github.com/OpenStitching/stitching/discussions/146

**Reply by neriniccolodei** [reply]
Thank you for your reply. 

I have followed the tutorial and it seems to provide the coordinates of the upper left corner of the images. Then, with size, I guess one could get the other corners by adding the width and height, simply. 

However, in my experience, printing the `final_corners` gives a result which is a tuple of negative numbers. Then, centring them returns a tuple of values which do not correspond to the corners I wish to retrieve.

I am using these two frames.
![frame001](https://github.com/OpenStitching/stitching/assets/62433425/2c1723b5-657c-4fb6-9963-d9eac3ca78a7)
![frame002](https://github.com/OpenStitching/stitching/assets/62433425/58cd128f-5be3-4364-ab20-59c87df08be0)

And I have these outputs:
`Final corners: [(-844, -632), (-776, -626)]
Final sizes: [(1627, 1090), (1614, 1078)]
Final centered corners: [(0, 0), (68, 6)]`

And the drawn corners are these, whilst the bottom right indicates the top left corner I retrieved by simply hovering my mouse pointer over it and reading the coordinates.
![Results](https://github.com/OpenStitching/stitching/assets/62433425/56f4b3ec-ae24-42b1-8d8a-f3c35c4dcf01)

Any further idea is very helpful.

Best regards

**Reply by lukasalexanderweber** [reply]
I don't have that much experience with the corners and sizes. The only thing I can suggest that you closely follow how points and sizes behave at warping and what's there role after blending.

**Reply by neriniccolodei** [reply]
Ok, will do. Thank you

**Reply by neriniccolodei** [reply]
Looking at the documentation a bit deeper, the corners and sizes are obtained by `low_corners, low_sizes = warper.warp_rois(low_sizes, cameras, camera_aspect)` (cell 44 in the tutorial). 
This uses the OpenCV `warpRoi`function, which _Returns projected image minimum bounding box_. ([https://docs.opencv.org/4.x/d5/d76/classcv_1_1PyRotationWarper.html#a3afece82103726fd89e9954015b18da3](url)

Therefore, I believe that corners and sizes just represent some sort of bounding box of the warped image. Not the warped image corners themselves. Would make sense because corners are not necessarily on the same horizontal or vertical lines after warping. And achieving them by adding some "size" would only result in a rectangle.

**Reply by lukasalexanderweber** [reply]
Ok I think I understand. Is there anything else you might want to try?

**Reply by neriniccolodei** [reply]
Honestly, I have no idea how to retrieve them precisely.

Starting with only one image, I've tried to apply the `PyRotationWarper.warpPoint` method to the four initial corners of image 1. That'd be top_left = (0,0); top_right = (width, 0), and so on. 
There's an [issue](https://github.com/opencv/opencv/issues/23450) about this method vs the `PyRotationWarper.warp` too. I am not sure this has been resolved. 

Anyway, retrieved top left corners in your tutorial, using weir images, are fine. I mean, the top left corner corresponds to the one you can roughly estimate just by looking at the warped masks in the cropper chapter of the tutorial. 
But using my images, they don't really make sense. I obtain negative numbers which means something outside of the image. And projecting them on the warped image doesn't put them in the right place, either. 
But yeah, so far I can't retrieve them and I'm stuck with this, but I am working on something else. 

I am now trying to make the stitching work using real-time video.

**Comment by artes14** [comment]
hey if you guys are still interested, I figured out how to get the top-left corners.
```
minx=min([a for a, b in final_corners])
miny=min([b for a, b in final_corners])
tmp = [(e1-minx, e2-miny) for e1, e2 in final_corners]
for point in tmp:
    panorama=cv.circle(panorama, point, 50, (0,255,0), 10)
```
it's every x, y coordinates minus the minimum of every x's and y's
